### ISSUE TITLE
Notice of Change of Trial Judge

### ISSUE THREAD

**Issue by cholly75** [ustaxcourt/ef-cms#8329]
As a clerk, so that I do not have to manually generate a Notice of Change of Trial Judge when a calendared trial session is reassigned, I need DAWSON to generate these automatically.

Occasionally, a trial session will need to be reassigned judges due to scheduling conflicts or absence. When this happens, clerks edit the Judge and the Chambers Phone Number on the Trial Session in DAWSON, and are then required to generate a Notice of Change of Trial Judge and serve it to the parties on each case in the affected trial session.

We would like DAWSON to automatically generate these notices using the data in the case, trial session, and the text in the attached documentation when the edits are made.

## Acceptance Criteria
* When a Trial Session's Judge value is changed on an Open trial session, generate the Notice of Change of Trial Judge for all Open cases in that session that are Proceeding Type of Regular according to the attached text and tokens:
[https://app.zenhub.com/files/152320868/1adc6ea4-3a76-4e12-a893-16238f3c53cf/download](https://app.zenhub.com/files/152320868/1adc6ea4-3a76-4e12-a893-16238f3c53cf/download)
* When a Trial Session's Judge value is changed on an Open trial session, generate the Notice of Change of Trial Judge for all Open cases in that session that are a Proceeding Type of Small according to the attached text and tokens:
[https://app.zenhub.com/files/152320868/62ea21f6-7fc6-4881-86df-5693a535bb18/download](https://app.zenhub.com/files/152320868/62ea21f6-7fc6-4881-86df-5693a535bb18/download)
* The tokens for both notices are defined below:
 * LOCATION = the Location data assigned to the trial session

 * PROCEEDING TYPE = the Proceeding Type of the case the notice is being generated for
 * TRIAL SESSION DATE = the date assigned to the trial session
 * PRIOR JUDGE TITLE = the title (Judge, Special Trial Judge, etc.) associated with the judge previously assigned to the trial session
 * PRIOR JUDGE FULL NAME = the full name of the judge previously assigned to the trial session
 * CURRENT JUDGE TITLE = the title (Judge, Special Trial Judge, etc.) associated with the judge currently assigned to the trial session
 * CURRENT JUDGE FULL NAME = the full name of the judge currently assigned to the trial session
 * CHAMBERS PHONE = the Chambers Phone Number currently assigned to the trial session
* The appropriate notices are generated upon saving changes to a trial session where the assigned judge value is replaced with a new judge
* Upon generating the notices:
 -PDFs are presented for printing and service for parties with paper service
 -Parties with electronic service receive service in the usual manner
 -The Notice of Change of Trial Judge is added to the docket record as event code "NOT" with the text "Notice of Change of Trial Judge" in the Filings and Proceedings column
 -The Notice of Change of Trial Judge is served on all parties.

## Tasks
- [x] Generate "NOT" when a calendared case has its judge updated for regular open cases - Rachel & Mrinalini (https://github.com/ustaxcourt/ef-cms/pull/2173)
  - [x] Create new Notice document 
  - [x] add docket entry for document
  - [x] Serve on generation
- [x] Generate "NOT" when a calendared case has its judge updated for small open cases - Rachel & Mrinalini
  - [x] Create new Notice document 
  - [x] add docket entry for document
  - [x] Serve on generation
- [x] Add new usecasehelper to appropriate appcontext  - Rachel & Mrinalini
- [x] Create interactor and infrastructure for document generator for NOT - Rachel & Mrinalini
- [x] Display all generated notices on paper service screen after trial session has been updated - Rachel & Mrinalini
- [x] Refactoring: DRY up setNoticeOfChange (trial judge and remote proceeding) files - Rachel & Mrinalini (https://github.com/ustaxcourt/ef-cms/pull/2177)
- [x] Refactoring: remove 'overwriteMain' - Mrinalini & Rachel
- [x] Add case caption and docket no. on NOT doc - Rachel & Mrinalini
- [x] Update font on NOT Doc - Rachel & Mrinalini
- [x] Standalone remote sessions should say 'remote' in place of location and proceeding type in the generated notice - Rosie and Rachel (https://github.com/ustaxcourt/ef-cms/pull/2182)
- [x] DOD: pull conditional check for if notice should be generated from `setNoticeOfChangeOfTrialJudge` useCaseHelper to `updateTrialSessionInteractor`

## Test Cases
* Clerk reassigns judge for Regular trial session; Notices of Change of Trial Judge are generated for all open cases on the session; text of notice matches AC provided; notices are presented for paper service and served electronically; notices are added to docket records appropriately
* Clerk reassigns judge for Small trial session; Notices of Change of Trial Judge are generated for all open cases on the session; text of notice matches AC provided; notices are presented for paper service and served electronically; notices are added to docket records appropriately
* Clerk reassigns judge for Hybrid trial session; Notices of Change of Trial Judge are generated for all open cases on the session; text of notice matches AC provided; notices are presented for paper service and served electronically; notices are added to docket records appropriately

## Story Definition of Ready (Created on 9/26/21)
The following criteria must be met in order for the user story to be picked up by the Flexion development team.
The user story must: 
- [ ]  Be immediately actionable for the development team
- [ ]  Be focused on users
- [ ]  Include a narrative. “As a ____, so that I can ______, I need _____.”
- [ ]  Include user-centric acceptance criteria that expresses the users' needs
- [ ]  Include user-centric testing scenarios
- [ ]  Include preconditions (story dependencies)
- [ ]  Not include design or technical solutions
- [ ]  Include non-standard testing, deployment, documentation, browser, mobile or security requirements, as well as any requirements for 3rd-party integrations

Process: Flexion developers and designers will test if the story meets acceptance criteria and test cases in Flexion dev and staging environments (“standard testing”). If additional acceptance criteria or testing scenarios are discovered while the story is in progress, a new story should be created, added to the backlog and prioritized by the product owner. 

## Definition of Done (Updated 10-6-21)
**Product Owner**
 - [ ]  Acceptance criteria have been met and validated on the Court's migration environment
 - [ ] Add scenario to testing document, if applicable (https://docs.google.com/spreadsheets/d/1FUHKC_YrT-PosaWD5gRVmsDzI1HS_U-8CyMIb-qX9EA/edit?usp=sharing)

**UX**
 - [ ] Business test scenarios have been refined to meet all acceptance criteria 
 - [ ] Usability has been validated
 - [ ] Wiki has been updated (if applicable) 
 - [ ] Story has been tested on a mobile device (for external users only)

**Engineering**
 - [x] Automated test scripts have been written
 - [x] Field level and page level validation errors (front-end and server-side) integrated and functioning
 - [x] Verify that language for docket record for internal users and external users is identical
 - [x] New screens have been added to pa11y scripts
 - [x] All new functionality verified to work with keyboard and macOS voiceover https://www.apple.com/voiceover/info/guide/_1124.html 
 - [x] READMEs, other appropriate docs, JSDocs and swagger/APIs fully updated
 - [x] UI should be touch optimized and responsive for external only (functions on supported mobile devices and optimized for screen sizes as required)
 - [x] Interactors should validate entities before calling persistence methods
 - [x] Code refactored for clarity and to remove any known technical debt
 - [x] Acceptance criteria for the story has been met
 - [x] If there are special instructions in order to deploy into the next environment, add them as a comment in the story
 - [x] Deployed to any Flexion environment for Engineering Review
 - [x] Deployed to the Court's migration environment

**Comment by ttlenard** [comment]
### **Test Cases**

**1. Clerk reassigns judge for Regular trial session; Notices of Change of Trial Judge are generated for all open cases on the session; text of notice matches AC provided; notices are presented for paper service and served electronically; notices are added to docket records appropriately**
- Log in as an Admissions Clerk
- Click on Trial Sessions menu
- On the Open tab, select a Regular trial session that is in the future by clicking on the City, State URL
- Click on the docket number to open a case 
- Click on Case information tab, and then on Parties tab (ensure that the petitioner currently has paper service)
- Click on the Edit link for a petitioner
- Add in an email address that you have access to for the petitioner so that you can receive service emails (if this is a new email address not yet associated with a case, you will need to verify it so that the email is not in a pending state)
- Next, log in as a Petitions Clerk
- Click on the Trial Sessions Menu
- Click on the Open Tab, and select the same regular trial session from the previous steps
- Click the edit button next to Session information
- Scroll down to Session Assignments area and change the current Judge to a new Judge's name
- Be sure to also update the Chambers Phone Number
- Scroll down and click Save

Expected Results

- [ ] Print for paper service screen appears and has appropriate notice available for all paper service petitioners for all open cases on the trial session.
- [ ] Notice includes the appropriate text for Regular trial session, and follows the format listed in the following example for Regular trial sessions: https://app.zenhub.com/files/152320868/1adc6ea4-3a76-4e12-a893-16238f3c53cf/download

- Click the done button
- You should be routed back to the Session information screen
- Click on the Docket number of one of the open cases on the trial session

Expected Results

- [ ] A Notice of Change of Trial Judge should be present on the Docket record
- [ ] Upon opening of the Notice, it should include the appropriate text for Regular trial session, and follows the format listed in the following example for Regular trial sessions: https://app.zenhub.com/files/152320868/1adc6ea4-3a76-4e12-a893-16238f3c53cf/download

- Go back to the Trial Session screen and click on the inactive Case tab
- Click on the Docket number of one of the docket numbers in the list

Expected Results

- [ ] No Notice of Change of Trial Judge will be on the docket record for any of the Inactive Cases in this list

Next, check the email that you associated with one of the petitioners on the case

Expected Results:

- [ ] You should have an email indicating that there was a Notice of Change of Trial Judge on the Case

**2. Clerk reassigns judge for Small trial session; Notices of Change of Trial Judge are generated for all open cases on the session; text of notice matches AC provided; notices are presented for paper service and served electronically; notices are added to docket records appropriately**
- Log in as an Admissions Clerk
- Click on Trial Sessions menu
- On the Open tab, select a Small trial session that is in the future by clicking on the City, State URL
- Click on the docket number to open a case 
- Click on Case information tab, and then on Parties tab (ensure that the petitioner currently has paper service)
- Click on the Edit link for a petitioner
- Add in an email address that you have access to for the petitioner so that you can receive service emails (if this is a new email address not yet associated with a case, you will need to verify it so that the email is not in a pending state)
- Next, log in as a Petitions Clerk
- Click on the Trial Sessions Menu
- Click on the Open Tab, and select the same Small trial session from the previous steps
- Click the edit button next to Session information
- Scroll down to Session Assignments area and change the current Judge to a new Judge's name
- Be sure to also update the Chambers Phone Number
- Scroll down and click Save

Expected Results

- [ ] Print for paper service screen appears and has appropriate notice available for all paper service petitioners for all open cases on the trial session.
- [ ] Notice includes the appropriate text for Small trial session, and follows the format listed in the following example for Small trial sessions: https://app.zenhub.com/files/152320868/62ea21f6-7fc6-4881-86df-5693a535bb18/download

- Click the done button
- You should be routed back to the Session information screen
- Click on the Docket number of one of the open cases on the trial session

Expected Results

- [ ] A Notice of Change of Trial Judge should be present on the Docket record
- [ ] Upon opening of the Notice, it should include the appropriate text for Small trial session, and follows the format listed in the following example for Small trial sessions: https://app.zenhub.com/files/152320868/62ea21f6-7fc6-4881-86df-5693a535bb18/download

- Go back to the Trial Session screen and click on the inactive Case tab
- Click on the Docket number of one of the docket numbers in the list

Expected Results

- [ ] No Notice of Change of Trial Judge will be on the docket record for any of the Inactive Cases in this list

Next, check the email that you associated with one of the petitioners on the case

Expected Results:

- [ ] You should have an email indicating that there was a Notice of Change of Trial Judge on the Case

**3. Clerk reassigns judge for Hybrid trial session; Notices of Change of Trial Judge are generated for all open cases on the session; text of notice matches AC provided; notices are presented for paper service and served electronically; notices are added to docket records appropriately**
- Log in as an Admissions Clerk
- Click on Trial Sessions menu
- On the Open tab, select a Hybrid trial session that is in the future by clicking on the City, State URL
- Click on the docket number to open a case that has a regular proceeding type
- Click on Case information tab, and then on Parties tab (ensure that the petitioner currently has paper service)
- Click on the Edit link for a petitioner
- Add in an email address that you have access to for the petitioner so that you can receive service emails (if this is a new email address not yet associated with a case, you will need to verify it so that the email is not in a pending state)
- Next, go back to the Hybrid trial session and click on a docket number to open a case that has a small proceeding type
- Click on Case information tab, and then on Parties tab (ensure that the petitioner currently has paper service)
- Click on the Edit link for a petitioner
- Add in an email address that you have access to for the petitioner so that you can receive service emails 
- Next, log in as a Petitions Clerk
- Click on the Trial Sessions Menu
- Click on the Open Tab, and select the same Hybrid trial session from the previous steps
- Click the edit button next to Session information
- Scroll down to Session Assignments area and change the current Judge to a new Judge's name
- Be sure to also update the Chambers Phone Number
- Scroll down and click Save

Expected Results

- [ ] Print for paper service screen appears and has appropriate Small proceeding type notice available for all paper service petitioners for all open small cases on the trial session.
- [ ] Print for paper service screen appears and has appropriate Regular proceeding type notice available for all paper service petitioners for all open regular cases on the trial session.
- [ ] For Small cases, Notice includes the appropriate text for Small trial session, and follows the format listed in the following example for Small trial sessions: https://app.zenhub.com/files/152320868/62ea21f6-7fc6-4881-86df-5693a535bb18/download
- [ ] For Regular cases, Notice includes the appropriate text for Regular trial session, and follows the format listed in the following example for Regular trial sessions: https://app.zenhub.com/files/152320868/1adc6ea4-3a76-4e12-a893-16238f3c53cf/download


- Click the done button
- You should be routed back to the Session information screen
- Click on the Docket number of one of the open Small proceeding type cases on the trial session

Expected Results

- [ ] A Notice of Change of Trial Judge should be present on the Docket record
- [ ] Upon opening of the Notice, it should include the appropriate text for Small trial session, and follows the format listed in the following example for Small trial sessions: https://app.zenhub.com/files/152320868/62ea21f6-7fc6-4881-86df-5693a535bb18/download
- [ ] Repeat for Regular case - ensure that the content of the Notice on the docket record follows the format listed in the following example for Regular trial sessions: https://app.zenhub.com/files/152320868/1adc6ea4-3a76-4e12-a893-16238f3c53cf/download

- Go back to the Trial Session screen and click on the inactive Case tab
- Click on the Docket number of one of the docket numbers in the list

Expected Results

- [ ] No Notice of Change of Trial Judge will be on the docket record for any of the Inactive Cases in this list

Next, check the email that you associated with one of the petitioners on the case

Expected Results:

- [ ] You should have 2 emails indicating that there was a Notice of Change of Trial Judge on the both Cases that you are associated with

**4. Be sure to test with Remote, Standalone Remote, and In Person Trial Sessions when testing the above test cases.**

**Comment by rachelschneiderman** [comment]
Waiting to hear from court if the language in the Notice of Change of Trial Judge for when a Remote trial session judge has changed is OK.

**Comment by rachelschneiderman** [comment]
This story is blocked as we are waiting on Docket - there is a potential change to the language in the generated notice.

**Comment by cholly75** [comment]
@rachelschneiderman - update to the language, which actually ends up being an answer to the previously asked question. Whaddya do.

In cases where the trial proceeding type of Standalone Remote has had the trial judge changed, instead of the opening sentence reading "The Standalone Remote, Remote trial session..." they would like it to read "The standalone remote trial session..."
